### PR TITLE
Support boringssl patches. Patch it to support gcc11.

### LINKED ISSUE #16: Build failure with gcc 11
With the (yet to be released) GCC 11 compilation of boringssl fails (for instance on Fedora 34) with the following error message:

```
In file included from /home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/bcm.c:38:
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/bn/asm/x86_64-gcc.c:427:51: error: argument 2 of type ‘const uint64_t[8]’ {aka ‘const long unsigned int[8]’} with mismatched bound [-Werror=array-parameter=]
  427 | void bn_sqr_comba8(BN_ULONG r[16], const BN_ULONG a[8]) {
      |                                    ~~~~~~~~~~~~~~~^~~~
In file included from /home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/bn/add.c:64,
                 from /home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/bcm.c:37:
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/bn/internal.h:292:51: note: previously declared as ‘const uint64_t[4]’ {aka ‘const long unsigned int[4]’}
  292 | void bn_sqr_comba8(BN_ULONG r[16], const BN_ULONG a[4]);
      |                                    ~~~~~~~~~~~~~~~^~~~
In file included from /home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/bcm.c:100:
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/sha/sha256.c: In function ‘SHA224_Final’:
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/sha/sha256.c:122:10: error: ‘SHA256_Final’ accessing 32 bytes in a region of size 28 [-Werror=stringop-overflow=]
  122 |   return SHA256_Final(out, ctx);
      |          ^~~~~~~~~~~~~~~~~~~~~~
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/sha/sha256.c:122:10: note: referencing argument 1 of type ‘uint8_t *’ {aka ‘unsigned char *’}
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/sha/sha256.c:172:20: note: in a call to function ‘SHA256_Final’
  172 | #define HASH_FINAL SHA256_Final
      |                    ^~~~~~~~~~~~
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/sha/../digest/md32_common.h:231:5: note: in expansion of macro ‘HASH_FINAL’
  231 | int HASH_FINAL(uint8_t out[HASH_DIGEST_LENGTH], HASH_CTX *c) {
      |     ^~~~~~~~~~
In file included from /home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/bcm.c:101:
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/sha/sha512.c: In function ‘SHA384_Final’:
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/sha/sha512.c:137:10: error: ‘SHA512_Final’ accessing 64 bytes in a region of size 48 [-Werror=stringop-overflow=]
  137 |   return SHA512_Final(out, sha);
      |          ^~~~~~~~~~~~~~~~~~~~~~
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/sha/sha512.c:137:10: note: referencing argument 1 of type ‘uint8_t *’ {aka ‘unsigned char *’}
/home/nmeum/android-tools/vendor/boringssl/crypto/fipsmodule/sha/sha512.c:196:5: note: in a call to function ‘SHA512_Final’
  196 | int SHA512_Final(uint8_t out[SHA512_DIGEST_LENGTH], SHA512_CTX *sha) {
      |     ^~~~~~~~~~~~
cc1: all warnings being treated as errors
```

This does not seem to be specific to this repository though as the same build failure occurs with [boringssl]( https://github.com/google/boringssl) git HEAD (6b48efac7b3b229c17cff55e5cfd9f9a0aea9b70). Regarding the specific warnings: The first seems to be mismatch between function prototype and function implementation. The second seems to be an array size mismatch which doesn't seem to be a real issue though as the passed `HASH_CTX` contains the amount of bytes which should actually be accessed. I think it would be best to report this to boringssl upstream. Unfortunately, their issue tracker requires a google account and I don't have one and don't plan on creating one.

### PR DESCRIPTION
**Pull request by vanaf**
Fixes #16.

### CODE REVIEW

File: patches/boringssl/0003-Disable-Werror.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index f58e853cd..b1fec9616 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -117,7 +117,7 @@ endif()
+ if(CMAKE_COMPILER_IS_GNUCXX OR CLANG)
+   # Note clang-cl is odd and sets both CLANG and MSVC. We base our configuration
+   # primarily on our normal Clang one.
+-  set(C_CXX_FLAGS "-Werror -Wformat=2 -Wsign-compare -Wmissing-field-initializers -Wwrite-strings -Wvla")
++  set(C_CXX_FLAGS "-Wformat=2 -Wsign-compare -Wmissing-field-initializers -Wwrite-strings -Wvla")
```
Comment (nmeum) [inline patches/boringssl/0003-Disable-Werror.patch:19]: Which warnings remain without this patch, i.e. which warnings cause the compilation to fail with `-Werror`? If possible, I would prefer it if boringssl would compile without modified `CFLAGS`. Also: If we disable `-Werror` anyhow, there is no need to fix the `-Wstringop-overflow` warning (which is what the second patch does) or is there?

Comment (vanaf) [reply patches/boringssl/0003-Disable-Werror.patch:19]: Only sha256.c warnings remains after 2 google CLs. See details in upstream bug: https://bugs.chromium.org/p/boringssl/issues/detail?id=402#c4
We apply 2 google patches to address the issues and apply our workaround patch to silence warnings of remaining issues

Comment (nmeum) [reply patches/boringssl/0003-Disable-Werror.patch:19]: Ah! In that case `-Wno-error=stringop-overflow` should be sufficient and preferable over disabling `-Werror` entirely.

Comment (nmeum) [reply patches/boringssl/0003-Disable-Werror.patch:19]: Ping, would be nice to include this as part of the upcoming 31.0.2 release (see #29) :)

**Comment by munix9** [comment]
I think for gcc11 this is needed
`-Wno-error=array-parameter -Wno-error=array-bounds -Wno-error=stringop-overflow`.

I have this as an example in the gh-actions for openSUSE Tumbleweed and Arch:base 
https://github.com/munix9/android-tools/actions/runs/898484500

I'm still preparing a PR for the gh-actions (a suggestion that might be built upon).

Note: This repository is considered a candidate for inclusion in openSUSE Factory/Tumbleweed
(https://build.opensuse.org/package/show/home:munix9:test/android-tools).

**Comment by vanaf** [comment]
There were some updates in upstream bug yesterday: https://bugs.chromium.org/p/boringssl/issues/detail?id=402#c7
Hopefully now it should build without patching cmakelists

**Comment by vanaf** [comment]
Created #34 for platform-tools-31.0.2 branch.

Verified package build well on Fedora 34.

**Comment by nmeum** [comment]
Closing this in favor of #34.

**Comment by nmeum** [comment]
Hm, actually it probably makes more sense to rebased this PR now that platform-tools-31.0.2 is merged, no? I think it's best to review PRs separately that's why I merged platform-tools-31.0.2 independently. Sorry, for the inconvenience!

File: vendor/CMakeLists.txt [hunk 2]
```diff
@@ -15,7 +14,8 @@ set(android-vendored
 	selinux
 	f2fs-tools
 	e2fsprogs
-	incremental_delivery)
+	incremental_delivery
+        boringssl)
```
Comment (nmeum) [inline vendor/CMakeLists.txt:18]: I know this is nitpicking but the indention seems wrong.

**Comment by nmeum** [comment]
Really an awful lot of patches but since these are all upstream patches we won't need them anymore after android updates its vendored boringssl version. I fixed the whitespace error, expanded the commit messages a bit, and merged this in  df7bb55525aa93831874b2be8d7b21a5101d1466.

**Pull request closed without merging**